src/cmd/services/m3coordinator/downsample/options.go

```diff
@@ -191,6 +207,287 @@ type Configuration struct {
 	BufferPastLimits []BufferPastLimitConfiguration `yaml:"bufferPastLimits"`
 }
 
+// RulesConfiguration is a set of rules configuration to use for downsampling.
+type RulesConfiguration struct {
+	// MappingRules are the mapping rules that sets retention and resolution
+	// for metrics given a filter to match metrics against.
+	MappingRules []MappingRuleConfiguration `yaml:"mappingRules"`
+
+	// RollupRules are the rollup rules that sets specific aggregations for sets
+	// of metrics given a filter to match metrics against.
+	RollupRules []RollupRuleConfiguration `yaml:"rollupRules"`
+}
+
+// MappingRuleConfiguration is a mapping rule configuration.
+type MappingRuleConfiguration struct {
+	// Filter is a string separated filter of labe name to label value
+	// glob patterns to filter the mapping rule to.
+	// e.g. "app:*nginx* foo:bar baz:qux*qaz*"
+	Filter string `yaml:"filter"`
+
+	// Aggregations is the aggregations to apply to the set of metrics.
+	// One of:
+	// - "Last"
+	// - "Min"
+	// - "Max"
+	// - "Mean"
+	// - "Median"
+	// - "Count"
+	// - "Sum"
+	// - "SumSq"
+	// - "Stdev"
+	// - "P10"
+	// - "P20"
+	// - "P30"
+	// - "P40"
+	// - "P50"
+	// - "P60"
+	// - "P70"
+	// - "P80"
+	// - "P90"
+	// - "P95"
+	// - "P99"
+	// - "P999"
+	// - "P9999"
+	Aggregations []aggregation.Type `yaml:"aggregations"`
+
+	// StoragePolicies is the retention/resolution storage policies to keep the
+	// matched metrics at.
+	StoragePolicies []StoragePolicyConfiguration `yaml:"storagePolicies"`
+
+	// Drop specifies to drop any metrics that match the filter rather than
+	// keeping them with a storage policy.
+	Drop bool `yaml:"drop"`
+
+	// Optional fields follows.
+
+	// Name is optional.
+	Name string `yaml:"name"`
+}
+
+// Rule returns the mapping rule for the mapping rule configuration.
+func (r MappingRuleConfiguration) Rule() (view.MappingRule, error) {
```

Comment:
nit: pull this out into separate file rather than in options?

Reply:
Yeah fair.

Reply:
Still might be a good idea (it may already be moved and my gh is cached or something though)